command/sync.go

```diff
@@ -0,0 +1,501 @@
+			defer stat.Collect(c.Command.FullName(), &err)()
+
+			return NewSync(c).Run(c)
+		},
+	}
+}
+
+type CommonObject struct {
+	src, dst *storage.Object
+}
+
+// Sync holds sync operation flags and states.
+type Sync struct {
+	src         string
+	dst         string
+	op          string
+	fullCommand string
+
+	// flags
+	delete   bool
+	sizeOnly bool
+
+	// s3 options
+	concurrency int
+	partSize    int64
+	storageOpts storage.Options
+}
+
+// NewSync creates Sync from cli.Context
+func NewSync(c *cli.Context) Sync {
+	return Sync{
+		src:         c.Args().Get(0),
+		dst:         c.Args().Get(1),
+		op:          c.Command.Name,
+		fullCommand: givenCommand(c),
+
+		// flags
+		delete:   c.Bool("delete"),
+		sizeOnly: c.Bool("size-only"),
+
+		// s3 options
+		partSize:    c.Int64("part-size") * megabytes,
+		concurrency: c.Int("concurrency"),
+		storageOpts: NewStorageOpts(c),
+	}
+}
+
+// Run starts copying given source objects to destination.
+func (s Sync) Run(c *cli.Context) error {
+	srcurl, err := url.New(s.src)
+	if err != nil {
+		return err
+	}
+
+	dsturl, err := url.New(s.dst)
+	if err != nil {
+		return err
+	}
+	sourceClient, err := storage.NewClient(c.Context, srcurl, s.storageOpts)
+	if err != nil {
+		return err
+	}
+
+	isBatch := srcurl.IsWildcard()
+	if !isBatch && !srcurl.IsRemote() {
+		obj, _ := sourceClient.Stat(c.Context, srcurl)
+		isBatch = obj != nil && obj.Type.IsDir()
+	}
+
+	sourceObjects, destObjects, err := s.GetSourceAndDestinationObjects(c.Context, srcurl, dsturl)
+	if err != nil {
+		printError(s.fullCommand, s.op, err)
+		return err
+	}
+	onlySource, onlyDest, commonObjects := CompareObjects(sourceObjects, destObjects)
+
+	// clear the arrays.
+	sourceObjects = nil
+	destObjects = nil
+
+	waiter := parallel.NewWaiter()
+
+	var (
+		merrorWaiter error
+		errDoneCh    = make(chan bool)
+	)
+
+	go func() {
+		defer close(errDoneCh)
+		for err := range waiter.Err() {
+			if strings.Contains(err.Error(), "too many open files") {
+				fmt.Println(strings.TrimSpace(fdlimitWarning))
+				fmt.Printf("ERROR %v\n", err)
+
+				os.Exit(1)
+			}
+			printError(s.fullCommand, s.op, err)
+			merrorWaiter = multierror.Append(merrorWaiter, err)
+		}
+	}()
+
+	strategy := strategy.New(s.sizeOnly)                                                             // create comparison strategy.
+	transferManager := transfer.NewManager(s.storageOpts, false, isBatch, s.concurrency, s.partSize) // create transfer manager
+	pipeReader, pipeWriter := io.Pipe()                                                              // create a reader, writer pipe to pass commands to run
+
+	var merrorPlan error // for errors in plan
+	// Create commands in background.
+	go func() {
+		s.PlanRun(c.Context, onlySource, onlyDest, commonObjects, dsturl, *transferManager, strategy, pipeWriter, &merrorPlan)
+	}()
+
+	// pass the reader to run
+	return multierror.Append(NewRun(c, pipeReader).Run(c), merrorPlan).ErrorOrNil()
```

Comment:
This line looks a bit compact. Let's remove `NewRun(...).Run()` call outside of this error handling.